internal/apps/dop/providers/rule/actions/pipeline/provider.go

```diff
@@ -0,0 +1,203 @@
+// Copyright (c) 2021 Terminus, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pipeline
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/erda-project/erda-infra/base/servicehub"
+	definitionpb "github.com/erda-project/erda-proto-go/core/pipeline/definition/pb"
+	sourcepb "github.com/erda-project/erda-proto-go/core/pipeline/source/pb"
+	"github.com/erda-project/erda-proto-go/dop/projectpipeline/pb"
+	"github.com/erda-project/erda/apistructs"
+	"github.com/erda-project/erda/bundle"
+	"github.com/erda-project/erda/internal/apps/dop/providers/projectpipeline"
+	"github.com/erda-project/erda/internal/apps/dop/services/apierrors"
+	"github.com/erda-project/erda/internal/apps/dop/services/branchrule"
+	"github.com/erda-project/erda/internal/apps/dop/services/pipeline"
+	"github.com/erda-project/erda/internal/pkg/diceworkspace"
+	"github.com/erda-project/erda/pkg/common/apis"
+)
+
+type provider struct {
+	bdl                *bundle.Bundle
+	ProjectPipelineSvc *projectpipeline.ProjectPipelineService `autowired:"erda.dop.projectpipeline.ProjectPipelineService"`
+	PipelineSource     sourcepb.SourceServiceServer            `autowired:"erda.core.pipeline.source.SourceService" required:"true"`
+	PipelineDefinition definitionpb.DefinitionServiceServer    `autowired:"erda.core.pipeline.definition.DefinitionService" required:"true"`
+	pipeline           *pipeline.Pipeline
+	branchRule         *branchrule.BranchRule
+}
+
+func (p *provider) WithPipelineSvc(svc *pipeline.Pipeline) {
+	p.pipeline = svc
+}
+
+func (p *provider) WithBranchRule(branchRule *branchrule.BranchRule) {
+	p.branchRule = branchRule
+}
+
+type Interface interface {
+	CreatePipeline(env map[string]interface{}) (string, error)
+	WithPipelineSvc(svc *pipeline.Pipeline)
+	WithBranchRule(branchRule *branchrule.BranchRule)
+}
+
+func (p *provider) Init(ctx servicehub.Context) error {
+	p.bdl = bundle.New(bundle.WithPipeline())
+	return nil
+}
+
+type PipelineConfig struct {
+	PipelineYml    string
+	PipelineYmlStr string
+	AppID          uint64
+	RefName        string
+	UserID         string
+	Path           string
+	FileName       string
+}
+
+func (p *provider) CreatePipeline(env map[string]interface{}) (string, error) {
+	gitPush, ok := env["git_push"]
```

Comment:
you should use `PipelineRuleEnv` here also.

Reply:
done